types/v2/json.go

```diff
@@ -0,0 +1,93 @@
+package v2
+
+import (
+	"encoding/json"
+	"io"
+	"strconv"
+
+	"github.com/honeycombio/honeycomb-opentracing-proxy/types"
+)
+
+type ZipkinJSONSpan struct {
+	TraceID        string                 `json:"traceId"`
+	Name           string                 `json:"name"`
+	ID             string                 `json:"id"`
+	ParentID       string                 `json:"parentId,omitempty"`
+	Kind           string                 `json:"kind,omitempty"`
+	LocalEndpoint  localEndpoint          `json:"localEndpoint,omitempty"`
+	RemoteEndpoint remoteEndpoint         `json:"remoteEndpoint,omitempty"`
+	Annotations    []*annotation          `json:"annotation,omitemptys"`
+	Tags           map[string]interface{} `json:"tags,omitempty"`
+	Debug          bool                   `json:"debug,omitempty"`
+	Timestamp      int64                  `json:"timestamp,omitempty"`
+	Duration       int64                  `json:"duration,omitempty"`
+}
+
+type annotation struct {
+	Timestamp int64           `json:"timestamp"`
+	Value     string          `json:"value"`
+	Host      *remoteEndpoint `json:"endpoint,omitempty"`
+}
+
+type localEndpoint struct {
+	Ipv4        string `json:"ipv4"`
+	Port        int    `json:"port"`
+	ServiceName string `json:"serviceName"`
+}
+
+type remoteEndpoint struct {
+	Ipv4 string `json:"ipv4"`
+	Port int    `json:"port"`
+}
+
+// DecodeJSON reads an array of JSON-encoded spans from an io.Reader, and
+// converts that array to a slice of Spans.
+func DecodeJSON(r io.Reader) ([]*types.Span, error) {
+	var jsonSpans []ZipkinJSONSpan
+	err := json.NewDecoder(r).Decode(&jsonSpans)
+	if err != nil {
+		return nil, err
+	}
+	spans := make([]*types.Span, len(jsonSpans))
+	for i, s := range jsonSpans {
+		spans[i] = convertJSONSpan(s)
+	}
+
+	return spans, nil
+}
+
+func convertJSONSpan(zs ZipkinJSONSpan) *types.Span {
+	traceIDAsInt, _ := strconv.ParseInt(zs.TraceID, 16, 64)
+	s := &types.Span{
+		CoreSpanMetadata: types.CoreSpanMetadata{
+			TraceID:      zs.TraceID,
+			TraceIDAsInt: traceIDAsInt,
+			Name:         zs.Name,
+			ID:           zs.ID,
+			ParentID:     zs.ParentID,
+			Debug:        zs.Debug,
+			DurationMs:   float64(zs.Duration) / 1000.,
```

Comment:
The decimal at the end seems out of place. Let's make it more clear that this is a float by adding a 0
```suggestion
			DurationMs:   float64(zs.Duration) / 1000.0,
```

Reply:
I guess that using time somewhere makes sense but I don't think I understood the comment, could you clarify? :)

CoreSpanMetadata.DurationMs is a float64, so I need to convert the end result. Or do you want me to change the type of CoreSpanMetadata.DurationMs to time.Duration? Although that would require changing a lot of other stuff too.

Reply:
The float conversion is fine - the issue is the use of `1000.` as a magic constant - I think we want `zs.Duration / float64(time.Microsecond)` here.

Reply:
Oh, alright. Done. Applied on both v1 and v2, since the old json used the 1000 constant.

Reply:
Yeah, I'm noticing some ones we probably should have been cleaner with 😆 Thanks for that